File: InterviewPratice/Javascript/sudoku2.js
const check = str => {
    let c = 0, chars = "";
    for(let i in str) {
        let ch = str.charAt(i);
        if(!chars.includes(ch) || ch == ".") { chars += ch; }
        else { c++; }
    }
    return c == 0;
};

const sudoku2 = grid => {
    let c = 0, str = "";
    for(let i = 0; i < 9; i++, str = "") {
        for(let j = 0; j < 9; j++) { str += grid[i][j]; }
        c = !check(str) ? c + 1 : c;
        str = "";
        for(let j = 0; j < 9; j++) { str += grid[j][i]; }
        c = !check(str) ? c + 1 : c;
    }
    for(let i = 0, x = 1; i < 3; i++, x += 3) {
        str = "";
        for(let j = 0, y = 1; j < 3; j++, y += 3, str = "") {
            str += grid[x][y];
            str += grid[x][y + 1];
            str += grid[x - 1][y + 1];
            str += grid[x + 1][y + 1];
            str += grid[x][y - 1];
            str += grid[x - 1][y - 1];
            str += grid[x + 1][y - 1];
            str += grid[x + 1][y];
            str += grid[x- 1][y];
            c = !check(str) ? c + 1 : c;
        }
    }
    return c == 0;
};

File: InterviewPratice/Java/Sudoku2.java
import java.util.Arrays;

final class Sudoku2 {
    
    boolean sudoku2(char[][] grid) {
        int[] row = new int[9], col = new int[9], box = new int[9];

        for(int i = 0; i < 9; i++) {
            for(int j = 0; j < 9; j++) {
                if(grid[i][j] != '.') row[Integer.parseInt(grid[i][j]+"") - 1]++;
                if(grid[j][i] != '.') col[Integer.parseInt(grid[j][i]+"") - 1]++;   
            }
            
            for(int j = 0; j < 9; j++) if(row[j] > 1 || col[j] > 1) return false;

            Arrays.fill(row, 0);
            Arrays.fill(col, 0);
        }
        
        for(int i = 1; i <= 9; i += 3) {
            for(int j = 1; j <= 9; j += 3) {
                if(grid[i][j] != '.')
                    box[Integer.parseInt(grid[i][j]+"") - 1]++;
                    
                if(grid[i][j + 1] != '.')
                    box[Integer.parseInt(grid[i][j + 1]+"") - 1]++;
                    
                if(grid[i - 1][j + 1] != '.')
                    box[Integer.parseInt(grid[i - 1][j + 1]+"") - 1]++;
                    
                if(grid[i + 1][j + 1] != '.')
                    box[Integer.parseInt(grid[i + 1][j + 1]+"") - 1]++;
                    
                if(grid[i][j - 1] != '.')
                    box[Integer.parseInt(grid[i][j - 1]+"") - 1]++;
                    
                if(grid[i - 1][j - 1] != '.')
                    box[Integer.parseInt(grid[i - 1][j - 1]+"") - 1]++;
                    
                if(grid[i + 1][j - 1] != '.')
                    box[Integer.parseInt(grid[i + 1][j - 1]+"") - 1]++;
                    
                if(grid[i + 1][j] != '.')
                    box[Integer.parseInt(grid[i + 1][j]+"") - 1]++;
                    
                if(grid[i - 1][j] != '.')
                    box[Integer.parseInt(grid[i - 1][j]+"") - 1]++;
                
                for(int e : box) if(e > 1) return false;
                
                Arrays.fill(box, 0);
            }
        }

        return true;
    }
}